### PR TITLE
feat!(install-local): run build inside an isolated env

### LINKED ISSUE #691: consider new approach to managing privileges, preferring root invocation
# Problem

I have noticed a particular design pattern common for AUR helpers, which appears also to have been adopted by the Pacstall utility.

The utility prefers to run as the regular user, and during its operation acquire permissions for elevated privileges, to be utilized when the built packages are ready for installation. 

The approach carries several limitations, including the following:

- The build work is performed and saved under the user directory, even though it has system-wide scope. Users may not want the work saved in their own space, and may benefit from invoking operations that share artifacts with operations invoked by other users.
- Users are accustomed to invoking package management tools with elevated privileges, and may include such invocations within operations already carrying  such privileges. 

# Proposal

An alternative approach would be the creation of a special user for Pacstall work operations, which carries file permissions to a designated system directory. The utility then may delegate work operations to a worker process with the reduced priviliges of this special user, and then, when the process completes, proceed with package installation.

Inverting the point at which privileges are changed, and in which direction, may  represent overall a design that is more elegant and better integrated.

# Notes

I have recently received my first exposure to Arch-family operating systems, and only a short time after learning about AUR, and wondering why the concept has not been adopted in other Linux families, came to learn of the Pacstall project. I am happy about this work, since I have been a long-time user of Ubuntu, and more recently Linux Mint, and have negotiated all the obstacles related to package management familiar to so many.

### PR DESCRIPTION
**Pull request by D-Brox**
## Purpose

`pacscripts` can be arbitrary code, and therefore can be unsafe.

This PR reduces the area of attack by sourcing the pacscripts  and running it's functions in an isolated environment by using bwrap. This is done by only giving read/write permissions inside the `$SRCDIR` source directory and the `$pkgdir` staging directory, while giving read-only to `/`, and no  internet access. After bwrap finishes executing, it will also kill with SIGKILL any child processes that are invoked and outlast it's parent.

## Approach

Created a new file called `bwrap.sh`, providing `safe_source()`, which sources a trimmed down env file created through bwrap, `bwrap_pkgver()` and `bwrap_function()`, which run the functions inside bwrap.

## Addendum

This is a breaking change, as `sudo` must be removed from every pacscript.

Closes #691

## Checklist

- [x] I confirm that I have read the [contributing guidelines](https://github.com/pacstall/pacstall/blob/develop/CONTRIBUTING.md), and this pull request is abiding by all the clauses stated in the guideline.

### CODE REVIEW

File: misc/scripts/install-local.sh [hunk 1]
```diff
@@ -1095,9 +1094,22 @@ function fail_out_functions() {
...
-    # NOTE: https://stackoverflow.com/a/29163890 (shorthand for 2>&1 |)
-    $func |& sudo tee "${LOGDIR}/$(printf '%(%Y-%m-%d_%T)T')-$name-$func.log" && return "${PIPESTATUS[0]}"
+    sudo bwrap --unshare-all --die-with-parent --new-session  \
+               --proc /proc --dev /dev --tmpfs /tmp --tmpfs /run \
+                --ro-bind / / --bind $STOWDIR $STOWDIR --bind $SRCDIR $SRCDIR \
+               --setenv LOGDIR "$LOGDIR" --setenv STGDIR "$STGDIR" \
+               --setenv STOWDIR "$STOWDIR" --setenv pkgdir "$pkgdir" \
+               "./$func.tmp"
+
+    return $?
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:1112]:
```suggestion
```
Not needed, it's implicit in bash when the last command runs.

**Review by Elsie19 — CHANGES_REQUESTED** [review]

File: misc/scripts/install-local.sh [hunk 2]
```diff
@@ -1095,9 +1094,19 @@ function fail_out_functions() {
 
 function run_function() {
     local func="$1"
+
+    echo "#! /bin/bash" | sudo tee "$func.tmp" > /dev/null
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:1098]:
```suggestion
    echo "#!/bin/bash" | sudo tee "$func.tmp" > /dev/null
```

It looks better

Comment (Elsie19) [inline misc/scripts/install-local.sh:1098]: We should also randomize the file name, for better security. Maybe with [mktemp](https://linux.die.net/man/1/mktemp)

**Review by Elsie19 — CHANGES_REQUESTED** [review]

File: misc/scripts/install-local.sh [hunk 3]
```diff
@@ -125,6 +126,20 @@ function log() {
     } | sudo tee "$METADIR/$name" > /dev/null
 }
 
+function run_pkgver() {
+    tmpfile=$(sudo mktemp -p $PWD)
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:130]:
```suggestion
    tmpfile="$(sudo mktemp -p "${PWD}")"
```

File: misc/scripts/install-local.sh [hunk 4]
```diff
@@ -125,6 +126,20 @@ function log() {
...
+    echo "#!/bin/bash" | sudo tee "$tmpfile" > /dev/null
+    echo "source ${envfile}" | sudo tee -a "$tmpfile" > /dev/null
+    echo "pkgver" | sudo tee -a "$tmpfile" > /dev/null
+    sudo chmod +rx "$tmpfile"
+
+    sudo env - bwrap --unshare-all --share-net --die-with-parent --new-session --ro-bind / /   \
+        --proc /proc --dev /dev --tmpfs /tmp --dev-bind /dev/null /dev/null \
+        --ro-bind $envfile $envfile --ro-bind $tmpfile $tmpfile \
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:138]:
```suggestion
        --ro-bind "$envfile" "$envfile" --ro-bind "$tmpfile" "$tmpfile" \
```

File: misc/scripts/install-local.sh [hunk 5]
```diff
@@ -722,7 +737,30 @@ mapfile -t FARCH < <(dpkg --print-foreign-architectures)
 export FARCH
 export CARCH="$(dpkg --print-architecture)"
 export DISTRO="$(set_distro)"
-if ! source "${pacfile}"; then
+
+function safe_source() {
+    export envfile=$(sudo mktemp -p $SRCDIR)
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:742]:
```suggestion
    export envfile="$(sudo mktemp -p "${SRCDIR}")"
```

File: misc/scripts/install-local.sh [hunk 6]
```diff
@@ -722,7 +737,30 @@ mapfile -t FARCH < <(dpkg --print-foreign-architectures)
...
+    sudo chmod +r "$envfile"
+
+    tmpfile=$(sudo mktemp -p $SRCDIR)
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:745]:
```suggestion
    tmpfile="$(sudo mktemp -p "${SRCDIR}")"
```

File: misc/scripts/install-local.sh [hunk 7]
```diff
@@ -722,7 +737,30 @@ mapfile -t FARCH < <(dpkg --print-foreign-architectures)
...
+    echo "source \"${pacfile}\"" | sudo tee -a "$tmpfile" > /dev/null
+    echo "/bin/env --unset LINES --unset COLUMNS --unset SHLVL --unset  _ \${OLD_ENV[@]} | \
+        /bin/sed -e 's/BASH_FUNC_//g;s/%%=() {/() {/g;s/^\(.*\)$/\1/g;\
+        s/^\(.[[:alnum:]_]*\)=\(.*\)/\1=\"\2/g;s/^\(.[[:alnum:]_]*\)=\(.*\)$/\1=\2\"/g' \
+        >> \"${envfile}\"" | sudo tee -a "$tmpfile" > /dev/null
+    sudo chmod +x "$tmpfile"
+
+    sudo env - bwrap --unshare-all --die-with-parent --new-session \
+        --proc /proc --dev /dev --tmpfs /tmp --tmpfs /run --dev-bind /dev/null /dev/null \
+        --ro-bind / / --bind $SRCDIR $SRCDIR  --setenv CARCH "$CARCH" --setenv DISTRO "$DISTRO" \
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:757]:
```suggestion
        --ro-bind / / --bind "$SRCDIR" "$SRCDIR" --setenv CARCH "$CARCH" --setenv DISTRO "$DISTRO" \
```

File: misc/scripts/install-local.sh [hunk 8]
```diff
@@ -722,7 +737,30 @@ mapfile -t FARCH < <(dpkg --print-foreign-architectures)
...
+        "$tmpfile"
+    sudo sed -i "s/BASH_FUNC_//g;s/%%=() {/() {/g;/^PPID=/d;/^EUID=/d" $envfile
+    sudo rm $tmpfile
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:760]:
```suggestion
    sudo rm "$tmpfile"
```

File: misc/scripts/install-local.sh [hunk 9]
```diff
@@ -722,7 +737,30 @@ mapfile -t FARCH < <(dpkg --print-foreign-architectures)
...
+}
+
+if ! safe_source || ! source $envfile; then
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:763]:
```suggestion
if ! safe_source || ! source "$envfile"; then
```

**Review by Elsie19 — CHANGES_REQUESTED** [review]

File: misc/scripts/install-local.sh [hunk 10]
```diff
@@ -49,8 +49,9 @@ function cleanup() {
     fi
     sudo rm -rf "${STOWDIR}/${name:-$PACKAGE}.deb"
     rm -f /tmp/pacstall-select-options
-    sudo rm -f $envfile
-    unset name repology pkgver epoch url depends makedepends breaks replace gives pkgdesc hash optdepends ppa arch maintainer pacdeps patch PACPATCH NOBUILDDEP provides incompatible optinstall epoch homepage backup pkgrel mask pac_functions repo envfile 2> /dev/null
+    sudo rm -f "$bwrapenv"
+    sudo rm -f "$safeenv"
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:53]:
```suggestion
    sudo rm -f "$bwrapenv" "$safeenv"
```

**Review by Elsie19 — CHANGES_REQUESTED** [review]

File: misc/scripts/install-local.sh [hunk 11]
```diff
@@ -1095,9 +1145,27 @@ function fail_out_functions() {
 
 function run_function() {
     local func="$1"
+    tmpfile=$(sudo mktemp -p $PWD)
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:1148]:
```suggestion
    tmpfile="$(sudo mktemp -p "${PWD}")"
```

File: misc/scripts/install-local.sh [hunk 12]
```diff
@@ -1095,9 +1145,27 @@ function fail_out_functions() {
...
     fancy_message sub "Running $func"
-    # NOTE: https://stackoverflow.com/a/29163890 (shorthand for 2>&1 |)
-    $func |& sudo tee "${LOGDIR}/$(printf '%(%Y-%m-%d_%T)T')-$name-$func.log" && return "${PIPESTATUS[0]}"
+    sudo bwrap --unshare-all --die-with-parent --new-session --ro-bind / / \
+        --proc /proc --dev /dev --tmpfs /tmp --tmpfs /run --dev-bind /dev/null /dev/null \
+        --bind "$STOWDIR" "$STOWDIR" --bind "$SRCDIR" "$SRCDIR" \
+        --setenv LOGDIR "$LOGDIR" --setenv STGDIR "$STGDIR" \
+        --setenv STOWDIR "$STOWDIR" --setenv pkgdir "$pkgdir" \
+        "$tmpfile"
+    sudo rm $tmpfile
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:1168]:
```suggestion
    sudo rm "$tmpfile"
```

**Review by Elsie19 — CHANGES_REQUESTED** [review]

**Comment by oklopfer** [comment]
@D-Brox can you fix these?
Screenshot 2023-11-11 at 2 09 43 PM

**Comment by D-Brox** [comment]
> @D-Brox can you fix these?
> Screenshot 2023-11-11 at 2 09 43 PM
> 

The first 2, yes.
The other 2 are probably false positives tho

**Comment by oklopfer** [comment]
@D-Brox how would this affect packages like [rhino-core](https://github.com/pacstall/pacstall-programs/blob/master/packages/rhino-core/rhino-core.pacscript) and [unicorn-desktop-git](https://github.com/pacstall/pacstall-programs/blob/master/packages/unicorn-desktop-git/unicorn-desktop-git.pacscript) that currently use `post_install` to place files + interact with the unisolated environment (temp packaging directories), like for config files and overwrites?

**Comment by D-Brox** [comment]
@oklopfer postinst runs inside apt, so no issue

**Comment by Elsie19** [comment]
> > @D-Brox can you fix these?
> > Screenshot 2023-11-11 at 2 09 43 PM
> 
> The first 2, yes. The other 2 are probably false positives tho

Can you fix the first two?

**Comment by oklopfer** [comment]
> @oklopfer `postinst` runs inside `apt`, so no issue

Yay

**Comment by oklopfer** [comment]
okay I did find a problem with this - certain pacscripts require a second connection to the internet in addition to the `url` variable, but running the builds in the isolated environment makes the second connections inaccessible. 

```bash
[+] INFO: Retrieving packages
Initializing download: https://github.com/hyprwm/xdg-desktop-portal-hyprland/archive/v1.2.5/v1.2.5.tar.gz
File size: unavailable
Opening output file v1.2.5.tar.gz
Server unsupported, starting from scratch with one connection.
Starting download

Connection 0 finished

Downloaded 44.6172 Kilobyte(s) in 0 second(s). (189.48 KB/s)
[+] INFO: Extracting v1.2.5.tar.gz
[+] INFO: Running functions
	[>] Running prepare
	[>] Running build
fatal: unable to access 'https://github.com/hyprwm/hyprland-protocols.git/': Could not resolve host: github.com

ERROR: Neither source directory 'build' nor build directory None contain a build file meson.build.
ninja: Entering directory `build'
ninja: error: loading 'build.ninja': No such file or directory
make: *** hyprland-share-picker: No such file or directory.  Stop.
	[!] ERROR: Could not build xdg-desktop-portal-hyprland properly
dpkg: warning: ignoring request to remove xdg-desktop-portal-hyprland which isn't installed
[+] INFO: Cleaning up
```

**Comment by D-Brox** [comment]
@oklopfer I can use a flag to share net if needed

**Comment by oklopfer** [comment]
@D-Brox
dew-it-galactic-republic

**Comment by Elsie19** [comment]
> @oklopfer I can use a flag to share net if needed

Maybe only for `prepare`, but `build` and `package` are prevented from accessing the internet.

File: misc/scripts/bwrap.sh [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    echo "readonly __OLD_ENV" | sudo tee -a "$tmpfile" > /dev/null
+    echo "source \"${input}\"" | sudo tee -a "$tmpfile" > /dev/null
+    # /bin/env returns variables and functions, with values, so we sed them out
+    echo "mapfile -t NEW_ENV < <(/bin/env -0 \${__OLD_ENV[@]} | \
+        sed -ze 's/BASH_FUNC_\(.*\)%%=\(.*\)$/\n/g;s/^\(.[[:alnum:]_]*\)=\(.*\)$/\1/g'|tr '\0' '\n')" | sudo tee -a "$tmpfile" > /dev/null
+    # The env sourced inside of bwrap should contain everything from the pacscripts
+    echo "declare -p \${NEW_ENV[@]} >> \"${bwrapenv}\"" | sudo tee -a "$tmpfile" > /dev/null
+    echo "declare -pf >> \"${bwrapenv}\"" | sudo tee -a "$tmpfile" > /dev/null
+    # The Pacstall env should only receive the bare minimum of information needed
+    echo "echo > \"${safeenv}\"" | sudo tee -a "$tmpfile" > /dev/null
```
Comment (Elsie19) [inline misc/scripts/bwrap.sh:46]: You can wrap all these inside `{}` and have just one `tee` call.

File: misc/scripts/bwrap.sh [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    "$tmpfile" && sudo rm "$tmpfile"
+    source "$safeenv"
+    sudo rm "$safeenv"
+}
+
+function bwrap_pkgver() {
+    tmpfile="$(sudo mktemp -p "${PWD}")"
+    echo "#!/bin/bash -e" | sudo tee "$tmpfile" > /dev/null
+    echo "source ${bwrapenv}" | sudo tee -a "$tmpfile" > /dev/null
+    echo "pkgver" | sudo tee -a "$tmpfile" > /dev/null
```
Comment (Elsie19) [inline misc/scripts/bwrap.sh:70]: Wrap in `{}` and use one `tee` call.

File: misc/scripts/bwrap.sh [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        --ro-bind "$bwrapenv" "$bwrapenv" --ro-bind "$tmpfile" "$tmpfile" \
+        "$tmpfile" && sudo rm "$tmpfile"
+}
+
+function bwrap_function() {
+    local func="$1"
+    tmpfile="$(sudo mktemp -p "${PWD}")"
+    echo "#!/bin/bash -a" | sudo tee "$tmpfile" > /dev/null
+    echo "mapfile -t OLD_ENV < <(compgen -A variable  -P \"--unset \")" | sudo tee -a "$tmpfile" > /dev/null
+    echo "source ${bwrapenv}" | sudo tee -a "$tmpfile" > /dev/null
```
Comment (Elsie19) [inline misc/scripts/bwrap.sh:84]: Same as above.

File: misc/scripts/install-local.sh [hunk 16]
```diff
@@ -1088,8 +1097,7 @@ else
 fi
 
 export srcdir="$PWD"
-sudo chown -R "$PACSTALL_USER":"$PACSTALL_USER" . 2> /dev/null
-
+sudo chown -R "root":"root" . 2> /dev/null
```
Comment (Elsie19) [inline misc/scripts/install-local.sh:1100]: These don't need quoting.

**Review by Elsie19 — CHANGES_REQUESTED** [review]

File: misc/scripts/bwrap.sh [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+    local input="${1}"
+    mkdir /tmp/pacstall 2>/dev/null
+    safeenv="$(sudo mktemp -p "${SRCDIR}")"
+    sudo chmod +r "$safeenv"
+    bwrapenv="$(sudo mktemp -t "${SRCDIR}/bwrapenv.XXXXXXXXXX")"
+    sudo chmod +r "$bwrapenv"
+    export bwrapenv
+
+    tmpfile="$(sudo mktemp -p "${SRCDIR}")"
+    echo "#!/bin/bash -ae" | sudo tee "$tmpfile" > /dev/null
```
Comment (Elsie19) [inline misc/scripts/bwrap.sh:35]: I don't like the `-e` flag. Causes *way* too many problems that all pacscripts are checked for either way.

**Review by Elsie19 — CHANGES_REQUESTED** [review]

**Comment by oklopfer** [comment]
this will never work in Docker containers, as they do not allow mounting of /proc in any capacity. `bwrap` must be abandoned in favor of a different wrapper. `fakeroot` or `firejail` may be alternatives.

**Comment by oklopfer** [comment]
complete in https://github.com/pacstall/pacstall/pull/1068

**Comment by Elsie19** [comment]
@D-Brox you have been replaced 😁

**Pull request closed without merging**